provider/datagen/src/transform/cldr/units/mod.rs

```diff
@@ -2,34 +2,137 @@
+                _ => ConstantType::Actual,
+            };
+
+            constants_map_in_str_form.insert(cons_name, (num, den, constant_type));
         }
 
-        let result = UnitsConstantsV1 {
-            constants_map: ZeroMap::from_iter(constants_map),
-        };
+        // This loop iterates over the constants, replacing any string values with their corresponding constant values.
+        let mut num_of_const_with_text;
+        loop {
+            num_of_const_with_text = 0;
+            let mut constants_with_constants_map_replaceable =
+                BTreeMap::<&str, (Vec<String>, Vec<String>, ConstantType)>::new();
+            for (cons_name, (num, den, constant_type)) in constants_map_in_str_form.iter() {
+                let mut temp_num = num.clone();
+                let mut temp_den = den.clone();
+                let mut temp_constant_type = *constant_type;
+
+                for i in 0..temp_num.len() {
+                    if !contains_alphabetic_chars(temp_num[i].as_str())
+                        || is_scientific_number(temp_num[i].as_str())
+                    {
+                        continue;
+                    }
+
+                    num_of_const_with_text += 1;
+                    if let Some((rnum, rden, rconstant_type)) =
+                        constants_map_in_str_form.get(temp_num[i].as_str())
+                    {
+                        temp_num.remove(i);
+                        // append the elements in rnum to num and rden to den
+                        temp_num.append(&mut rnum.clone());
+                        temp_den.append(&mut rden.clone());
+
+                        if *rconstant_type == ConstantType::Approximate {
+                            temp_constant_type = ConstantType::Approximate;
+                        }
+                    }
+                }
+
+                for i in 0..temp_den.len() {
+                    if !contains_alphabetic_chars(temp_den[i].as_str())
+                        || is_scientific_number(temp_den[i].as_str())
+                    {
+                        continue;
+                    }
+
+                    num_of_const_with_text += 1;
+                    if let Some((rnum, rden, rconstant_type)) =
+                        constants_map_in_str_form.get(temp_den[i].as_str())
+                    {
+                        temp_den.remove(i);
+                        // append the elements in rnum to den and rden to num
+                        temp_num.append(&mut rden.clone());
+                        temp_den.append(&mut rnum.clone());
+
+                        if *rconstant_type == ConstantType::Approximate {
+                            temp_constant_type = ConstantType::Approximate;
+                        }
+                    }
+                }
+
+                constants_with_constants_map_replaceable
+                    .insert(cons_name, (temp_num, temp_den, temp_constant_type));
+            }
```

Comment:
If there are cycles in the constants, this loop will not terminate. Guard against this.

Reply:
There is no specific number of cycles; however, the loop must terminate, or there is a problem with the CLDR data.

Reply:
Yes, and datagen should terminate with an error if it's called with invalid data.

Reply:
I have made the maximum depth is 10 for now.

Reply:
I don't think that's a good solution. There's nothing that says the data cannot use more than 10 steps, and if it does in the future we'll have a problem if there's a hardcoded limit. Instead of setting a limit, you should *detect* if there's any cycle.

Reply:
Yes, but if the number of cycle are more than 10, you will get an error. Because one of the constants will have something like "ft5_to_m5". Then, there will be error that "ft5_to_m5" is not going to parse.

However, I added a cleaner error. good catch.

Reply:
No. Needing 11 cycles to fully resolve all variables is perfectly fine, we cannot set an arbitrary limit at 10. In fact, any number is perfectly fine, it just cannot be infinity. 

You also cannot tell users to go into the code and change the value, imagine if software you were using did that.

Please detect the loop.

Reply:
This is now a catch-22 situation. If we detect the loop (which is currently set to 2) and the CLDR increases the depth in the future, we must adjust the depth. Otherwise, if we leave the loop infinite with the break condition, the loop will continue indefinitely if the CLDR mistakenly has a loop in its constants definition.

Reply:
You can detect if you're in a loop by keeping track of nodes you've seen, i.e. constants you have expanded.

Reply:
solved